[C++ API] torch::nn::FractionalMaxPool{2,3}d module and functional

yf225 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29933 [C++ API] torch::nn::FractionalMaxPool{2,3}d module and functional**

Differential Revision: [D18548174](https://our.internmc.facebook.com/intern/diff/D18548174)

Commit:
[C++ API] torch::nn::FractionalMaxPool{2,3}d module and functional

Commit:
Update on "[C++ API] torch::nn::FractionalMaxPool{2,3}d module and functional"

[C++ API] torch::nn::FractionalMaxPool{2,3}d module and functional

gh-metadata: pytorch pytorch 29933 gh/yf225/77/head

kostmo commented:
## CircleCI build failures summary
As of commit 8ef33d4:
* **2/2** failures introduced in this PR
* **0/2** recognized as flaky



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=8ef33d4dfbe4fb3a8e852f143d494b756f79d532):

| Job | Step | Log excerpt |
| --- | --- | --- |
| pytorch_linux_backward_compatibility_check_test | Test | `The PR is introducing backward incompatible changes` |
| pytorch_linux_xenial_py3_clang5_asan_test | Test | `RuntimeError: ` |


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

Commit:
Update on "[C++ API] torch::nn::FractionalMaxPool{2,3}d module and functional"

[C++ API] torch::nn::FractionalMaxPool{2,3}d module and functional

gh-metadata: pytorch pytorch 29933 gh/yf225/77/head

Commit:
Update on "[C++ API] torch::nn::FractionalMaxPool{2,3}d module and functional"

[C++ API] torch::nn::FractionalMaxPool{2,3}d module and functional

gh-metadata: pytorch pytorch 29933 gh/yf225/77/head

pbelevich commented on torch/csrc/api/src/nn/modules/pooling.cpp:
```diff
@@ -273,6 +273,91 @@ template class MaxUnpoolImpl<3, MaxUnpool3dImpl>;
+
+void FractionalMaxPool2dImpl::reset() {
+  _random_samples = register_buffer("_random_samples", options._random_samples());
+  if (options.output_size() == c10::nullopt && options.output_ratio() == c10::nullopt) {
```
I don't insist, but WDYT about this?
```suggestion
  if (!options.output_size() && !options.output_ratio()) {
```

yf225 replied:
I think either way works - I was using `== c10::nullopt` because it looks more similar to Python version and also enforces the type of `options.output_size()` more strongly, in the sense that if we ever change the type of `options.output_size()` to not be `c10::optional<...>`, `options.output_size() == c10::nullopt` would fail to compile, while `!options.output_size()` might still compile (because the new type might also overload the `bool()` operator). In such a scenario, failing to compile is preferred over being able to compile, because we won't be sure if the new type's `bool()` operator is checking the right thing that we care about at the original call site.

pbelevich commented on torch/csrc/api/src/nn/modules/pooling.cpp:
```diff
@@ -273,6 +273,91 @@ template class MaxUnpoolImpl<3, MaxUnpool3dImpl>;
+  if (options.output_size() != c10::nullopt && options.output_ratio() != c10::nullopt) {
+    TORCH_CHECK(false, "only one of output_size and output_ratio may be specified");
+  }
+  if (options.output_ratio() != c10::nullopt) {
```
WDYT?
```suggestion
  if (options.output_ratio()) {
```

yf225 replied:
Replied in an earlier comment.

pbelevich commented on torch/csrc/api/include/torch/nn/functional/pooling.h:
```diff
@@ -494,6 +494,125 @@ inline Tensor max_unpool3d(const Tensor& input, const Tensor& indices,
+
+  c10::optional<ExpandingArray<2>> output_size_ = output_size;
+  if (output_size_ == c10::nullopt) {
+    TORCH_INTERNAL_ASSERT(output_ratio != c10::nullopt);
```
WDYT?
```suggestion
    TORCH_INTERNAL_ASSERT(output_ratio);
```

yf225 replied:
Replied in an earlier comment.

pbelevich commented on torch/csrc/api/include/torch/nn/functional/pooling.h:
```diff
@@ -494,6 +494,125 @@ inline Tensor max_unpool3d(const Tensor& input, const Tensor& indices,
+
+  c10::optional<ExpandingArray<3>> output_size_ = output_size;
+  if (output_size_ == c10::nullopt) {
+    TORCH_INTERNAL_ASSERT(output_ratio != c10::nullopt);
```
WDYT?
```suggestion
    TORCH_INTERNAL_ASSERT(output_ratio);
```

yf225 replied:
Replied in an earlier comment.

pbelevich commented on torch/csrc/api/src/nn/modules/pooling.cpp:
```diff
@@ -273,6 +273,91 @@ template class MaxUnpoolImpl<3, MaxUnpool3dImpl>;
+      "FractionalMaxPool3d requires specifying either ",
+      "an output size, or a pooling ratio");
+  }
+  if (options.output_size() != c10::nullopt && options.output_ratio() != c10::nullopt) {
```
```suggestion
  if (options.output_size() && options.output_ratio()) {
```

yf225 replied:
Replied in an earlier comment.

pbelevich commented on torch/csrc/api/src/nn/modules/pooling.cpp:
```diff
@@ -273,6 +273,91 @@ template class MaxUnpoolImpl<3, MaxUnpool3dImpl>;
+
+void FractionalMaxPool3dImpl::reset() {
+  _random_samples = register_buffer("_random_samples", options._random_samples());
+  if (options.output_size() == c10::nullopt && options.output_ratio() == c10::nullopt) {
```
```suggestion
  if (!options.output_size() && !options.output_ratio()) {
```

yf225 replied:
Replied in an earlier comment.

pbelevich commented on torch/csrc/api/include/torch/nn/functional/pooling.h:
```diff
@@ -494,6 +494,125 @@ inline Tensor max_unpool3d(const Tensor& input, const Tensor& indices,
+    const c10::optional<ExpandingArray<2>>& output_size,
+    const c10::optional<ExpandingArray<2>>& output_ratio,
+    const Tensor& _random_samples) {
+  if (output_size == c10::nullopt && output_ratio == c10::nullopt) {
```
WDYT?
```suggestion
  if (!output_size && !output_ratio) {
```

yf225 replied:
Replied in an earlier comment.

pbelevich commented on torch/csrc/api/include/torch/nn/functional/pooling.h:
```diff
@@ -494,6 +494,125 @@ inline Tensor max_unpool3d(const Tensor& input, const Tensor& indices,
+  }
+
+  c10::optional<ExpandingArray<2>> output_size_ = output_size;
+  if (output_size_ == c10::nullopt) {
```
WDYT?
```suggestion
  if (!output_size_) {
```

yf225 replied:
Replied in an earlier comment.

pbelevich commented on torch/csrc/api/src/nn/modules/pooling.cpp:
```diff
@@ -273,6 +273,91 @@ template class MaxUnpoolImpl<3, MaxUnpool3dImpl>;
+Tensor FractionalMaxPool2dImpl::forward(const Tensor& input) {
+  return F::detail::fractional_max_pool2d(
+            input, options.kernel_size(), options.output_size(), options.output_ratio(),
+            _random_samples=_random_samples);
```
I didn't get it, maybe just
```suggestion
            _random_samples);
```

yf225 replied:
I was surprised this doesn't throw a compile error :/ Fixed.

pbelevich commented on torch/csrc/api/src/nn/modules/pooling.cpp:
```diff
@@ -273,6 +273,91 @@ template class MaxUnpoolImpl<3, MaxUnpool3dImpl>;
+std::tuple<Tensor, Tensor> FractionalMaxPool2dImpl::forward_with_indices(const Tensor& input) {
+  return F::detail::fractional_max_pool2d_with_indices(
+            input, options.kernel_size(), options.output_size(), options.output_ratio(),
+            _random_samples=_random_samples);
```
I didn't get it, maybe just
```suggestion
            _random_samples);
```

yf225 replied:
Fixed

pbelevich commented on torch/csrc/api/src/nn/modules/pooling.cpp:
```diff
@@ -273,6 +273,91 @@ template class MaxUnpoolImpl<3, MaxUnpool3dImpl>;
+Tensor FractionalMaxPool3dImpl::forward(const Tensor& input) {
+  return F::detail::fractional_max_pool3d(
+            input, options.kernel_size(), options.output_size(), options.output_ratio(),
+            _random_samples=_random_samples);
```
I didn't get it, maybe just
```suggestion
            _random_samples);
```

yf225 replied:
Fixed

pbelevich commented on torch/csrc/api/src/nn/modules/pooling.cpp:
```diff
@@ -273,6 +273,91 @@ template class MaxUnpoolImpl<3, MaxUnpool3dImpl>;
+Tensor FractionalMaxPool3dImpl::forward(const Tensor& input) {
+  return F::detail::fractional_max_pool3d(
+            input, options.kernel_size(), options.output_size(), options.output_ratio(),
+            _random_samples=_random_samples);
```
maybe
```suggestion
            _random_samples);
```

yf225 replied:
Fixed

pbelevich commented on torch/csrc/api/src/nn/modules/pooling.cpp:
```diff
@@ -273,6 +273,91 @@ template class MaxUnpoolImpl<3, MaxUnpool3dImpl>;
+std::tuple<Tensor, Tensor> FractionalMaxPool3dImpl::forward_with_indices(const Tensor& input) {
+  return F::detail::fractional_max_pool3d_with_indices(
+            input, options.kernel_size(), options.output_size(), options.output_ratio(),
+            _random_samples=_random_samples);
```
maybe
```suggestion
            _random_samples);
```

yf225 replied:
Fixed

Commit:
Update on "[C++ API] torch::nn::FractionalMaxPool{2,3}d module and functional"

[C++ API] torch::nn::FractionalMaxPool{2,3}d module and functional

gh-metadata: pytorch pytorch 29933 gh/yf225/77/head

Commit:
Update on "[C++ API] torch::nn::FractionalMaxPool{2,3}d module and functional"

[C++ API] torch::nn::FractionalMaxPool{2,3}d module and functional

gh-metadata: pytorch pytorch 29933 gh/yf225/77/head

Commit:
Update on "[C++ API] torch::nn::FractionalMaxPool{2,3}d module and functional"

[C++ API] torch::nn::FractionalMaxPool{2,3}d module and functional

gh-metadata: pytorch pytorch 29933 gh/yf225/77/head

Commit:
Update on "[C++ API] torch::nn::FractionalMaxPool{2,3}d module and functional"

[C++ API] torch::nn::FractionalMaxPool{2,3}d module and functional

gh-metadata: pytorch pytorch 29933 gh/yf225/77/head

Commit:
Update on "[C++ API] torch::nn::FractionalMaxPool{2,3}d module and functional"

[C++ API] torch::nn::FractionalMaxPool{2,3}d module and functional

gh-metadata: pytorch pytorch 29933 gh/yf225/77/head

Commit:
Update on "[C++ API] torch::nn::FractionalMaxPool{2,3}d module and functional"

[C++ API] torch::nn::FractionalMaxPool{2,3}d module and functional

gh-metadata: pytorch pytorch 29933 gh/yf225/77/head

Commit:
Update on "[C++ API] torch::nn::FractionalMaxPool{2,3}d module and functional"

[C++ API] torch::nn::FractionalMaxPool{2,3}d module and functional

gh-metadata: pytorch pytorch 29933 gh/yf225/77/head

Commit:
Update on "[C++ API] torch::nn::FractionalMaxPool{2,3}d module and functional"

[C++ API] torch::nn::FractionalMaxPool{2,3}d module and functional

gh-metadata: pytorch pytorch 29933 gh/yf225/77/head

Commit:
Update on "[C++ API] torch::nn::FractionalMaxPool{2,3}d module and functional"

[C++ API] torch::nn::FractionalMaxPool{2,3}d module and functional

gh-metadata: pytorch pytorch 29933 gh/yf225/77/head

Commit:
Update on "[C++ API] torch::nn::FractionalMaxPool{2,3}d module and functional"

[C++ API] torch::nn::FractionalMaxPool{2,3}d module and functional

gh-metadata: pytorch pytorch 29933 gh/yf225/77/head

yf225 commented on torch/csrc/api/include/torch/nn/modules/pooling.h:
```diff
@@ -397,6 +397,74 @@ class TORCH_API MaxUnpool3dImpl : public MaxUnpoolImpl<3, MaxUnpool3dImpl> {
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ FractionalMaxPool3d ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+/// Applies fractional maxpool over a 3-D input.
+/// See https://pytorch.org/docs/master/nn.html#torch.nn.FractionalMaxPool3d to learn
```
Removed
